src/engine.rs

```diff
@@ -0,0 +1,187 @@
+use crate::bytes::{CharIterator, Confidence, Encoding};
+use crate::dns::ResolveType;
+use crate::html5::parser::document::{Document, DocumentBuilder, DocumentHandle};
+use crate::html5::parser::Html5Parser;
+use crate::net::errors::Error;
+use crate::net::http::headers::Headers;
+use crate::net::http::request::Request;
+use crate::net::http::response::Response;
+use crate::timing::{Timing, TimingTable};
+use crate::types::ParseError;
+use cookie::CookieJar;
+use core::fmt::Debug;
+use core::str::FromStr;
+use reqwest::header::HeaderName;
+use reqwest::Method;
+use url::Url;
+
+/// Response that is returned from the fetch function
+pub struct FetchResponse {
+    /// Request that has been send
+    pub request: Request,
+    /// Response that has been received
+    pub response: Response,
+    /// Document tree that is made from the response
+    pub document: DocumentHandle,
+    /// Parse errors that occurred while parsing the document tree
+    pub parse_errors: Vec<ParseError>,
+    /// Rendertree that is generated from the document tree and css tree
+    pub render_tree: String,
+    /// Timing table that contains all the timings
+    pub timings: TimingTable,
+}
+
+impl Debug for FetchResponse {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        writeln!(f, "Request:")?;
+        writeln!(f, "{}", self.request)?;
+        writeln!(f, "Response:")?;
+        writeln!(f, "{}", self.response)?;
+        writeln!(f, "Document tree:")?;
+        writeln!(f, "{}", self.document)?;
+        writeln!(f, "Parse errors:")?;
+        for error in &self.parse_errors {
+            writeln!(f, "  ({}:{}) {}", error.line, error.col, error.message)?;
+        }
+        writeln!(f, "Render tree:")?;
+        writeln!(f, "{}", self.render_tree)?;
+        writeln!(f, "Timings:")?;
+        writeln!(f, "{}", self.timings)?;
+
+        Ok(())
+    }
+}
+
+fn fetch_url(
+    method: &str,
+    url: &str,
+    headers: Headers,
+    cookies: CookieJar,
+) -> Result<FetchResponse, Error> {
```

Comment:
```suggestion
) -> Result<FetchResponse> {
```

To get this to work, you'd need to use the `Result` type from `crate::types`, and you might need to translate the error.